src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,164 @@
+            throw new IndexOutOfBoundsException(
+                    String.format("Index %d out of bounds. List size: %d", index, size)
+            );
+        }
+        addBefore(value, index);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> currentNode = findByIndex(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToRemove = findByIndex(index);
+        unlink(nodeToRemove);
+        return nodeToRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.value == object
+                    || (currentNode.value != null
+                    && currentNode.value.equals(object))) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findByIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(
+                    String.format("Index %d out of bounds. List size: %d", index, size)
+            );
+        }
+        Node<T> currentNode;
+        if (index < (size >> 1)) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size; i > (index + 1); i--) {
+                currentNode = currentNode.previous;
+            }
+        }
+        return currentNode;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.next == null) {
+            if (node.previous != null) {
+                node.previous.next = null;
+            }
+            tail = node.previous;
+        } else if (node.previous == null) {
+            node.next.previous = null;
+            head = node.next;
+        } else {
+            node.next.previous = node.previous;
+            node.previous.next = node.next;
+        }
+        size--;
+    }
+
+    private void addAtFirstPosition(T value) {
+        Node<T> firstElement = head;
+        Node<T> newNode = new Node<>(null, value, firstElement);
+        head = newNode;
+        if (firstElement == null) {
+            tail = newNode;
```

Comment:
this if clause is never happen, because of code above

```
 public void add(T value, int index) {
        if (index == size) {
            add(value);
            return;
        }
```